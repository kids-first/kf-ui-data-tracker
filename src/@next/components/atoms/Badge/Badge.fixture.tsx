import React from 'react';
import {useSelect} from 'react-cosmos/fixture';

import {Badge} from './Badge';

const Fixture = () => {
    const [color] = useSelect('color', {
        options: [
            'gray',
            'red',
            'yellow',
            'green',
            'blue',
            'indigo',
            'purple',
            'pink',
        ],
    });

    return (
        <div className="p-8">
            <Badge color={color}>5 new messages</Badge>
        </div>
    );
};

export default Fixture;
