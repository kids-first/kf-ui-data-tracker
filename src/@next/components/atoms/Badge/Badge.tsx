import React from 'react';

export interface IBadgeProps {
    color?:
        | 'gray'
        | 'red'
        | 'yellow'
        | 'green'
        | 'blue'
        | 'indigo'
        | 'purple'
        | 'pink';
    children: React.ReactNode;
}

export const Badge = ({color, children}: IBadgeProps) => (
    <span
        className={`inline-flex items-center px-2.5 py-0.5 rounded-full text-xs font-medium bg-${color}-100 text-${color}-800`}
    >
        {children}
    </span>
);

export default Badge;
