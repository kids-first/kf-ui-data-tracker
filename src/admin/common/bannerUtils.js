const isActive = banner => {
  /* 
    Banner is active if it is enabled && 
    start date <= today && end date is after today 
    or end date is null
    */
  if (!banner.enabled) {
    return false;
  }
  // Set today to localtime midnight
  const today = new Date().setHours(0, 0, 0, 0);
  if (banner.startDate && banner.endDate) {
    return (
      Date.parse(banner.startDate) <= today &&
      Date.parse(banner.endDate) > today
    );
  } else if (banner.startDate) {
    return Date.parse(banner.startDate) <= today;
  } else {
    return false;
  }
};

const activeDisplayBanners = banners => {
  /*
   * Display the 2 newest (by start date) banners
   * */
  return banners
    .map(banner => banner.node)
    .filter(banner => isActive(banner))
    .sort((a, b) => a.startDate - b.startDate)
    .slice(0, 2);
};

export default activeDisplayBanners;
