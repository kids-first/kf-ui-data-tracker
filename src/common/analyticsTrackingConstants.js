/**
 * This file contains string constants for Amplitude eventTypes and
 * eventProperties in the format of:
 * {
 *  [<scope_string>] : {
 *     scope: <scope_string>,
 *     [<event_type_string>] : <event_type_string>,
 *      ...event_key_value_pairs
 *  }
 * }
 *
 * let scopeEvent = scope.event
 * --> scope_event
 */

/**
 * reflects a list of keys to key:value pairs
 * @param {array} keys
 */
function keyMirror(keys) {
  keys = Array.isArray(keys) ? keys : Object.keys(keys);
  var mirror = {};
  keys.forEach(v => (mirror[v] = v));
  return mirror;
}

const events = eventsList => keyMirror(eventsList);

/**
 * creates a proxied object with getters that will
 * return the event name with the scope name appended (<scope_name>_<event_name>)
 *
 * This is used to sanitize and keep all events and scopes consistent
 *
 * @param {string} name scopeName
 * @param {array} eventsList array of event name strings
 * @return {
 *   <SCOPE_NAME>: Proxy {
 *     scope:  <SCOPE_NAME>,
 *     <EVENT_NAME>: <EVENT_NAME>,
 *
 *      get [<EVENT_NAME>](){
 *        return this.<SCOPE_NAME>+_+<EVENT_NAME>
 *      }
 *  }
 * }
 */
const scope = (name, eventsList) => {
  let scopeOriginal = {
    scope: name.trim().toUpperCase(),
    ...events(
      eventsList
        .filter(x => typeof x === 'string')
        .map(e =>
          e
            .trim()
            .replace(' ', '_')
            .toUpperCase(),
        ),
    ),
  };
  /**
   * @see https://stackoverflow.com/a/7891968
   */
  let scopeProxy = new Proxy(scopeOriginal, {
    get(target, name, receiver) {
      let rv = Reflect.get(target, name, receiver);
      if (typeof rv === 'string' && name !== 'scope') {
        rv = Reflect.get(target, 'scope', receiver) + '_' + rv;
      }
      return rv;
    },
  });

  return {
    [name.trim().toUpperCase()]: scopeProxy,
  };
};

const MOUSE_EVENTS = ['HOVER', 'CLICK'];
const DOCUMENT_EVENTS = ['UPLOAD', 'DOWNLOAD', 'DELETE', 'EDIT'];
const INPUT_EVENTS = ['_BLUR', '_FOCUS', '_CHANGE'];
const MODAL_EVENTS = ['MODAL__OPEN', 'MODAL__CLOSE'];

/**
 * analytics tracking constants object
 */
const analyticsTrackingConstants = {
  ...scope('INPUT', ['TEXT', ...INPUT_EVENTS]),
  ...scope('MOUSE', MOUSE_EVENTS),
  ...scope('AUTH', ['LOGIN', 'LOGOUT']),
  ...scope('PAGE', ['VIEW']),
  ...scope('STUDY_LIST_', ['VIEW_TOGGLE', 'SEARCH', 'ADD_STUDY']),
  ...scope('STUDY_TABLE_', MOUSE_EVENTS),
  ...scope('STUDY_CARD_', ['TOGGLE_DETAIL', ...MOUSE_EVENTS]),
  ...scope('STUDY_INFO_', DOCUMENT_EVENTS),
  ...scope('UPLOAD_WIZARD_', ['STEP', 'CLOSE']),
  ...scope('DOCUMENT_', [...DOCUMENT_EVENTS, ...MODAL_EVENTS]),
  ...scope('DOCUMENT_INFO_', DOCUMENT_EVENTS),
  ...scope('DOCUMENT_VERSION_', [
    ...DOCUMENT_EVENTS,
    ...MOUSE_EVENTS,
    ...MODAL_EVENTS,
  ]),
  ...scope('NEW_DOCUMENT_', DOCUMENT_EVENTS),
  ...scope('NEW_VERSION_', [...DOCUMENT_EVENTS, ...MOUSE_EVENTS]),
  ...scope('TOOLTIP', MOUSE_EVENTS),
  ...scope('AVATAR_', MOUSE_EVENTS),
};

export default analyticsTrackingConstants;
