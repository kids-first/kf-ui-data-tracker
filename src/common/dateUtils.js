// Reformat time to the "DD MMM YYYY" format like "12 Aug 2019"
export const longDate = date => {
  var mydate = new Date(date);
  var month = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sept',
    'Oct',
    'Nov',
    'Dec',
  ][mydate.getMonth()];
  var str = mydate.getUTCDate() + ' ' + month + ' ' + mydate.getFullYear();
  return str;
};

export const toYearMonthDay = isoDate => {
  /*
   * Convert ISO8601 UTC datetime string to local datetime and format as
   * yyyy-mm-dd date string.
   * HTML input type=date requires the yyyy-mm-dd format
   */
  let date = '';
  if (isoDate) {
    date = new Date(isoDate);
    let month = '' + (date.getMonth() + 1);
    let day = '' + date.getDate();
    let year = date.getFullYear();

    if (month.length < 2) month = '0' + month;
    if (day.length < 2) day = '0' + day;

    date = [year, month, day].join('-');
  }
  return date;
};

// Reformat time to the date and time format
export const dateTime = date => {
  var mydate = new Date(date);
  var month = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sept',
    'Oct',
    'Nov',
    'Dec',
  ][mydate.getMonth()];
  const weekday = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  var str =
    weekday[mydate.getDay()] +
    ', ' +
    month +
    ' ' +
    mydate.getUTCDate() +
    ', ' +
    mydate.getFullYear() +
    ', ' +
    mydate.toLocaleTimeString([], {hour: '2-digit', minute: '2-digit'});
  return str;
};
