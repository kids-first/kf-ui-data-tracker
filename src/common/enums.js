export const projectOptions = [
  {key: 'DEL', value: 'DEL', text: 'Delivery', icon: 'paper plane outline'},
  {key: 'HAR', value: 'HAR', text: 'Analysis', icon: 'sliders horizontal'},
  {key: 'RES', value: 'RES', text: 'Research', icon: 'flask'},
];

// Different kinds of workflows Cavatica projects
export const workflowOptions = [
  {key: 'bwa_mem', value: 'bwa_mem', text: 'bwa_mem'},
  {key: 'bwa_mem_bqsr', value: 'bwa_mem_bqsr', text: 'bwa_mem_bqsr'},
  {key: 'star_2_pass', value: 'star_2_pass', text: 'star_2_pass'},
  {
    key: 'gatk_haplotypecaller',
    value: 'gatk_haplotypecaller',
    text: 'GATK Haplotypecaller',
  },
  {
    key: 'gatk_genotypgvcf',
    value: 'gatk_genotypgvcf',
    text: 'GATK Genotypgvcf',
  },
  {
    key: 'gatk_genotypegvcf_vqsr',
    value: 'gatk_genotypegvcf_vqsr',
    text: 'GATK GenotypeGVCF VQSR',
  },
  {
    key: 'strelka2_somatic_mode',
    value: 'strelka2_somatic_mode',
    text: 'Strelka2 Somatic Mode',
  },
  {
    key: 'mutect2_somatic_mode',
    value: 'mutect2_somatic_mode',
    text: 'MuTect2 Somatic Mode',
  },
  {
    key: 'mutect2_tumor_only_mode',
    value: 'mutect2_tumor_only_mode',
    text: 'MuTect2 Tumor Only Mode',
  },
  {
    key: 'vardict_single_sample_mode',
    value: 'vardict_single_sample_mode',
    text: 'VarDict Single Sample Mode',
  },
  {
    key: 'vardict_paired_sample_mode',
    value: 'vardict_paired_sample_mode',
    text: 'VarDict Paired Sample Mode',
  },
  {
    key: 'control_freec_somatic_mode',
    value: 'control_freec_somatic_mode',
    text: 'Control FREEC Somatic Mode',
  },
  {
    key: 'control_freec_germline_mode',
    value: 'control_freec_germline_mode',
    text: 'Control FREEC Germline Mode',
  },
  {
    key: 'stringtie_expression',
    value: 'stringtie_expression',
    text: 'StringTie Expression',
  },
  {key: 'manta_somatic', value: 'manta_somatic', text: 'Manta Somatic'},
  {key: 'manta_germline', value: 'manta_germline', text: 'Manta Germline'},
  {key: 'lumpy_somatic', value: 'lumpy_somatic', text: 'LUMPY Somatic'},
  {key: 'lumpy_germline', value: 'lumpy_germline', text: 'LUMPY Germline'},
  {key: 'rsem', value: 'rsem', text: 'RSEM'},
  {key: 'kallisto', value: 'kallisto', text: 'Kallisto'},
  {key: 'star_fusion', value: 'star_fusion', text: 'Star Fusion'},
  {key: 'arriba', value: 'arriba', text: 'Arriba'},
  {key: 'peddy', value: 'peddy', text: 'peddy'},
];

// Styling for different event types
export const eventType = {
  SF_CRE: {
    title: 'Study File Created',
    iconName: 'add',
    iconColor: 'green',
  },
  SF_UPD: {
    title: 'Study File Updated',
    iconName: 'refresh',
    iconColor: 'yellow',
  },
  SF_DEL: {
    title: 'Study File Deleted',
    iconName: 'delete',
    iconColor: 'red',
  },
  FV_CRE: {
    title: 'File Version Created',
    iconName: 'add',
    iconColor: 'green',
  },
  FV_UPD: {
    title: 'File Version Updated',
    iconName: 'refresh',
    iconColor: 'yellow',
  },
  SD_CRE: {title: 'Study Created', iconName: 'add', iconColor: 'green'},
  SD_UPD: {
    title: 'Study Updated',
    iconName: 'refresh',
    iconColor: 'yellow',
  },
  PR_CRE: {
    title: 'Project Created',
    iconName: 'add',
    iconColor: 'green',
  },
  PR_UPD: {
    title: 'Project Updated',
    iconName: 'refresh',
    iconColor: 'green',
  },
  PR_DEL: {
    title: 'Project Deleted',
    iconName: 'delete',
    iconColor: 'red',
  },
  PR_LIN: {
    title: 'Project Linked',
    iconName: 'linkify',
    iconColor: 'green',
  },
  PR_UNL: {
    title: 'Project Unlinked',
    iconName: 'unlinkify',
    iconColor: 'red',
  },
  OTH: {title: 'Other', iconName: 'question', iconColor: 'blue'},
};

// Store file type title, description and icon
export const fileTypeDetail = {
  SHM: {
    icon: 'shipping',
    title: 'Shipping Manifest',
    description: 'File type description goes here...',
  },
  CLN: {
    icon: 'hospital',
    title: 'Clinical/Phenotype Data',
    description: 'File type description goes here...',
  },
  SEQ: {
    icon: 'dna',
    title: 'Sequencing Manifest',
    description: 'File type description goes here...',
  },
  OTH: {
    icon: 'question',
    title: 'Other',
    description: 'File type description goes here...',
  },
};

// Store version state title and color
export const versionState = {
  PEN: {title: 'Pending review', labelColor: 'orange'},
  APP: {title: 'Approved', labelColor: 'teal'},
  CHN: {title: 'Changes needed', labelColor: 'red'},
  PRC: {title: 'Processed', labelColor: 'blue'},
  UPD: {title: 'Updated', labelColor: 'blue'},
};

// Store label colors for system version footer
export const systemEnvColors = {
  local: 'orange',
  dev: 'yellow',
  prd: 'green',
  qa: 'blue',
};
