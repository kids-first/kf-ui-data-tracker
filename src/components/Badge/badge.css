.Badge {
  @inherit: .max-w-full, .self-center, .m-4, .px-16, .py-0, .rounded-full,
    .text-center, .text-xs, .text-white, .leading-tight, .whitespace-no-wrap;
}

.Badge--loading {
  @inherit: .bg-lightGrey, .max-w-full, .self-center, .m-4, .px-16, .py-0,
    .rounded-full, .text-center, .text-xs, .text-white, .leading-tight,
    .whitespace-no-wrap;
}

.Badge--invalid {
  @inherit: .bg-darkGrey, .max-w-full, .self-center, .m-4, .px-16, .py-0,
    .rounded-full, .text-center, .text-xs, .text-white, .leading-tight,
    .whitespace-no-wrap;
}
