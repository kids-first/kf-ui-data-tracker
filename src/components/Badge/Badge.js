import React from 'react';
import PropTypes from 'prop-types';
import {versionState} from '../../common/fileUtils';
import {Label} from 'semantic-ui-react';
/**
 * Displays badge describing the file state
 */
const Badge = ({state, size}) => (
  <Label
    basic
    size={size}
    color={versionState[state] ? versionState[state].labelColor : 'grey'}
  >
    {state && versionState[state] ? versionState[state].title : 'Invalid'}
  </Label>
);

Badge.propTypes = {
  /** Badge state. */
  state: PropTypes.string,
  /** Badge size. */
  size: PropTypes.oneOf([
    'mini',
    'tiny',
    'small',
    'medium',
    'large',
    'big',
    'huge',
    'massive',
  ]),
};

Badge.defaultProps = {
  state: null,
  size: 'tiny',
};

export default Badge;
