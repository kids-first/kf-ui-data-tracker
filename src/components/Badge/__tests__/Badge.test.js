import React from 'react';
import {render, cleanup} from '@testing-library/react';
import {versionState} from '../../../common/enums';
import Badge from '../Badge';

afterEach(cleanup);
it('renders badges with all available state', async () => {
  const stateList = Object.keys(versionState);
  var i;
  for (i in stateList) {
    const {container} = render(<Badge state={stateList[i]} />);
    expect(container).toMatchSnapshot();
  }
});

it('renders badges in different sizes', async () => {
  const sizeList = [
    'mini',
    'tiny',
    'small',
    'medium',
    'large',
    'big',
    'huge',
    'massive',
  ];
  var i;
  for (i in sizeList) {
    const {container} = render(<Badge state="PEN" size={sizeList[i]} />);
    expect(container).toMatchSnapshot();
  }
});

it('renders badge with customized icon', async () => {
  const {container} = render(<Badge icon="star outline" />);
  expect(container).toMatchSnapshot();
});

it('renders badge on loading mode', async () => {
  const {container} = render(<Badge loading state={null} />);
  expect(container).toMatchSnapshot();
});

it('renders badge with missing values for size and state', async () => {
  const {container} = render(<Badge />);
  expect(container).toMatchSnapshot();
});
