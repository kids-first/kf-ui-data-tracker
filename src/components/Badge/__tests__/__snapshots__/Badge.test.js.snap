// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders badge on loading mode 1`] = `
<div>
  <div
    class="ui grey tiny basic label"
  >
    Invalid
  </div>
</div>
`;

exports[`renders badges with all available state 1`] = `
<div>
  <div
    class="ui orange tiny basic label"
  >
    Pending review
  </div>
</div>
`;

exports[`renders badges with all available state 2`] = `
<div>
  <div
    class="ui teal tiny basic label"
  >
    Approved
  </div>
</div>
`;

exports[`renders badges with all available state 3`] = `
<div>
  <div
    class="ui red tiny basic label"
  >
    Changes needed
  </div>
</div>
`;

exports[`renders badges with all available state 4`] = `
<div>
  <div
    class="ui blue tiny basic label"
  >
    Processed
  </div>
</div>
`;
